### PR TITLE
[Dataviews] Table layout: Ensure focus is not lost on interaction

### LINKED ISSUE #57341: [DataViews] When we hide fields from the columns in `table` view, there is a focus loss
When columns are hidden from the column header menu, the menu is necessarily removed from the DOM, which leaves focus at the whim of the user agent. We need to be intentional about focus movement in this case.

Additionally, there are others areas of interaction that can lead to focus loss. When tables are sorted, filtered, or paginated, new data can be fetched. If this takes time to come back, the header is currently replaced with a "no results" message, until the fetch is complete.

### PR DESCRIPTION
**Pull request by andrewhayward**
Resolves #57341.

## What?

This PR does ~three~ two relatively small things:
1. It changes the table header rendering behaviour, to always be present even when no data is available.
2. It makes sure that focus is suitably moved when a column is hidden from its header menu.
3. ~It persists stale data, to maintain visual consistency, which is rendered and marked as `inert` while fresh data is being fetched (sorting, pagination, etc).~

## Why?

When tables are sorted, filtered, or paginated, new data can be fetched. If this takes time to come back, the header is currently replaced with a "no results" message, until the fetch is complete. With the table header remaining in place, focus within header menus is persisted.

Separately, when columns are hidden from the column header menu, the menu is necessarily removed from the DOM, which leaves focus at the whim of the user agent. We need to be intentional about focus movement in this case.

## How?

1. The table header rendering is moved outside of the `hasData` check, so that it is always rendered, regardless of whether data exists.
2. A map is maintained of column headers, and suitable fallbacks, so that focus can be appropriately moved if a column is hidden.
3. ~A reference is maintained of previously rendered data, which is used when no current data is available.~

## Testing Instructions

1. Open the "Manage all pages" view in the new data views manager
2. Sort the data using a column header menu
   * Confirm that sorting different columns still works
   * Confirm that focus remains in the header menu
3. Paginate the data using the pagination controls
   * Confirm that data is still correctly updated
4. Filter the data (using a column header menu or filter controls)
   * Confirm that data is still correctly updated
   * Confirm that a 'No results' message is shown if no data matches the filter
5. Filter the data from the column header menu, to leave no results
   * Confirm that focus remains in the header menu
6. Remove a column using a column header menu
   * Confirm that focus is moved to the header menu in the previous column
   * Note: if removing the first column, focus should move to the next column instead

### Testing Instructions for Keyboard

The same interactions should be validated using the keyboard – focus should not be lost at any point.

## Screenshots or screencast

The table layout with no data – the header row is persisted.

The DataViews Pages view, using the table layout. The data is filtered, with no results, and the message 'No results' is shown under the header row.

Column header removal, showing focus being explicitly moved to a different column.

https://github.com/WordPress/gutenberg/assets/159848/b2aeb3e0-e345-4e7d-98d2-9d850585c314

### CODE REVIEW

**Comment by jameskoster** [comment]
Very nice, do you think it's worth rebasing to check this against the latest work around dropdown menus?

**Comment by andrewhayward** [comment]
> ...do you think it's worth rebasing to check this against the latest work around dropdown menus?

Yes. I thought I was up to date when I cut the branch, but evidently not! Rebased now, things still seem to work without issue.

https://github.com/WordPress/gutenberg/assets/159848/80f0a6e1-7e2c-4c11-a6eb-018ba9e82f9d

File: packages/dataviews/src/view-table.js [hunk 1]
```diff
@@ -122,12 +132,14 @@ function HeaderMenu( { field, view, onChangeView } ) {
 					<DropdownMenuItem
 						prefix={ <Icon icon={ unseen } /> }
 						onClick={ () => {
-							onChangeView( {
-								...view,
-								hiddenFields: view.hiddenFields.concat(
-									field.id
-								),
-							} );
+							if ( ( onHide && onHide( field ) ) ?? true ) {
```
Comment (mirka) [inline packages/dataviews/src/view-table.js]: I found it hard to grok the intent of this conditional since the possible/expected return values of `onHide()` are unclear and undocumented. And the name `onHide` suggests that it would normally return undefined, not a boolean.

I have a feeling it would be more clearly expressed as an early return, but in any case it still isn't intuitive that something called `onHide()` would return a `false`. Was the intent something else?

```js
if ( onHide?.( field ) === false ) return;
onChangeView( /* ... */ );
```

File: packages/dataviews/src/view-table.js [hunk 2]
```diff
@@ -291,54 +316,64 @@ function ViewTable( {
 			)
 	);
 	const shownData = useAsyncList( data );
-	const usedData = deferredRendering ? shownData : data;
+	const usefulData = data.length ? data : shownData;
+	const deferredData = deferredRendering ? shownData : usefulData;
+	const staleData = shownData.length ? shownData : previousData.current;
+	const usedData =
+		isLoading && ! deferredData.length ? staleData : deferredData;
```
Comment (mirka) [inline packages/dataviews/src/view-table.js]: This section was a bit hard to follow for me. I had some suggestions to make it clearer, based on what I understood, but I might be misunderstanding some things.

1. `shownData` does not seem like an accurate name, as it may or may not be the data shown. For naming consistency within this section of logic, we should probably stick to the adjective "deferred".
2. The `usefulData` variable name wasn't sufficiently informative for me, but the logic was actually easier to understand without that intermediate variable. So maybe we can merge that logic into the next conditional.
3. I am not exactly sure why we need to fall back to the async data when `data.length` is zero. Wouldn't the async data have a zero length in that case as well? (A code comment would help if there's a reason.)
4. The `deferredData` variable name in `deferredData = deferredRendering ? shownData : usefulData` also seemed inaccurate, as it is sometimes not the async data.

```diff
diff --git a/packages/dataviews/src/view-table.js b/packages/dataviews/src/view-table.js
index cd7ac7cc39..e89e9707ae 100644
--- a/packages/dataviews/src/view-table.js
+++ b/packages/dataviews/src/view-table.js
@@ -315,12 +315,12 @@ function ViewTable( {
 				field.id
 			)
 	);
-	const shownData = useAsyncList( data );
-	const usefulData = data.length ? data : shownData;
-	const deferredData = deferredRendering ? shownData : usefulData;
-	const staleData = shownData.length ? shownData : previousData.current;
+	const deferredData = useAsyncList( data );
+	const maybeDeferredData =
+		deferredRendering || ! data.length ? deferredData : data;
+	const staleData = deferredData.length ? deferredData : previousData.current;
 	const usedData =
-		isLoading && ! deferredData.length ? staleData : deferredData;
+		isLoading && ! maybeDeferredData.length ? staleData : maybeDeferredData;
 	previousData.current = usedData;
 	const hasData = !! usedData?.length;
```

File: packages/dataviews/src/view-table.js [hunk 3]
```diff
@@ -291,54 +316,64 @@ function ViewTable( {
...
+									sortValues[ view.sort.direction ]
+								}
+								scope="col"
+							>
+								<HeaderMenu
+									ref={ ( node ) =>
+										( headerMenuRefs.current[ field.id ] = {
+											node,
+											fallback:
+												visibleFields[ index - 1 ]?.id,
```
Comment (mirka) [inline packages/dataviews/src/view-table.js]: Should this fall back to `index + 1` when the current index is 0?

File: packages/dataviews/src/view-table.js [hunk 4]
```diff
@@ -283,6 +295,19 @@ function ViewTable( {
...
+	const onHide = ( field ) => {
+		const hidden = headerMenuRefs.current[ field.id ];
+		const fallback = headerMenuRefs.current[ hidden.fallback ];
+		// These stacked microtask callbacks are required to
+		// make sure the node is available to be focused
+		queueMicrotask( () => {
+			queueMicrotask( () => {
+				if ( fallback?.node ) fallback.node.focus();
+			} );
+		} );
```
Comment (mirka) [inline packages/dataviews/src/view-table.js]: I am uneasy with this use of `queueMicrotask`, as it relies on an incidental fact that we currently need to wait two re-renders, which I'm not sure is not guaranteed to be true forever.

Have we considered other ways to achieve this, ideally in a more explicit manner? Here is one way I can think of, but there may be better ways:

```js
	const [ headerFieldToFocus, setHeaderFieldToFocus ] = useState( null );
	const onHide = ( field ) => {
		const hidden = headerMenuRefs.current[ field.id ];
		const fallback = headerMenuRefs.current[ hidden.fallback ];
		if ( fallback?.node ) {
			setHeaderFieldToFocus( fallback.node );
		}
	};

	headerFieldToFocus?.focus();
	if (
		headerFieldToFocus?.ownerDocument.activeElement === headerFieldToFocus
	) {
		setHeaderFieldToFocus( null );
	}
```

**Review by mirka — COMMENTED** [review]
Great improvement!

The focus behavior in particular seems like a great thing to have automated tests for. There doesn't seem to be any tests for this component yet so I'm not sure if we're waiting for things to stabilize more, but in the long run it'd be great to have test coverage since this will be a very complicated component.

File: packages/dataviews/src/view-table.js [hunk 5]
```diff
@@ -366,14 +401,13 @@ function ViewTable( {
...
-				</table>
-			) }
-			{ ! hasData && (
-				<div className="dataviews-no-results">
-					<p>{ __( 'No results' ) }</p>
-				</div>
-			) }
+				</tbody>
+			</table>
+			<div className="dataviews-no-results" id={ tableNoticeId }>
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js]: Given this component is no longer about "no results" but holds any message related to the table (loading, no results), the class could be renamed to `dataviews-table-notice`/`dataviews-table-caption` or something along those lines.

File: packages/dataviews/src/view-table.js [hunk 6]
```diff
@@ -291,54 +316,64 @@ function ViewTable( {
...
-										maxWidth: field.maxWidth || undefined,
-									} }
-									data-field-id={ field.id }
-									aria-sort={
-										view.sort?.field === field.id &&
-										sortValues[ view.sort.direction ]
+			<table
+				className="dataviews-table-view"
+				aria-busy={ isLoading || undefined }
+				aria-describedby={ tableNoticeId }
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js:349]: Nice touch here https://www.w3.org/WAI/tutorials/tables/caption-summary/#using-aria-describedby-to-provide-a-table-summary

File: packages/dataviews/src/view-table.js [hunk 7]
```diff
@@ -291,54 +316,64 @@ function ViewTable( {
...
-										minWidth: field.minWidth || undefined,
-										maxWidth: field.maxWidth || undefined,
-									} }
-									data-field-id={ field.id }
-									aria-sort={
-										view.sort?.field === field.id &&
-										sortValues[ view.sort.direction ]
+			<table
+				className="dataviews-table-view"
+				aria-busy={ isLoading || undefined }
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js]: Why fallback [aria-busy](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-busy) to `undefined` if it's false?

File: packages/dataviews/src/view-table.js [hunk 8]
```diff
@@ -283,6 +295,19 @@ function ViewTable( {
 	isLoading = false,
 	deferredRendering,
 } ) {
+	const headerMenuRefs = useRef( {} );
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js]: We could use a `Map` and, in the ref callback, make sure the ref is removed when the node is also removed from the DOM:

```js
ref={ ( node ) => {
  if ( node ) {											 
    headerMenuRefs.current.set( field.id, { ... } );
  } else {
    headerMenuRefs.current.delete( field.id );
  }
} }
```

File: packages/dataviews/src/view-table.js [hunk 9]
```diff
@@ -39,7 +45,10 @@ const sanitizeOperators = ( field ) => {
 	);
 };
 
-function HeaderMenu( { field, view, onChangeView } ) {
+const HeaderMenu = forwardRef( function HeaderMenu(
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js:55]: nit pick: what about passing an unnamed arrow function to `forwardRef`?

```js
const HeaderMenu = forwardRef( ( props, ref ) => { ... } );
```

Comment (oandregal) [reply packages/dataviews/src/view-table.js:55]: Actually, it's best to leave as it is: React devtools report it as named and not anonymous.

File: packages/dataviews/src/view-table.js (hunk 2 above)
Comment (oandregal) [reply packages/dataviews/src/view-table.js]: Would you be open to extract this to a separate PR? It seems this would be useful to any view type (not only table), and is unrelated to focus control. We could lift this logic to `DataViews` and only pass `data` to the view component.

Comment (oandregal) [reply packages/dataviews/src/view-table.js]: While looking at other things, I had a related thought: have you looked at leveraging how the data retrieval hook ([useEntityRecords](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/page-pages/index.js#L183)) works? It maintains a cache, and the stale `data` is sometimes maintained (and some other times returns `null`) while it goes through the lifecycle.

**Comment by oandregal** [comment]
Captura de ecrã 2023-12-29, às 14 30 01

Just a wrangling update: I've changed the `Type` label from `enhancement` to `experimental`.

The `Type` labels (and others) are used to generate the Gutenberg changelog for releases (see [example](https://make.wordpress.org/core/2023/12/20/whats-new-in-gutenberg-17-3-20-december/) under "Experiments > DataViews"). Right now, `DataViews` is an experiment, so if we all tag our PRs as `[Type] Experimental` it makes life easier for the person that wrangles the Gutenberg release (all PRs grouped and don't pollute other sections and no manual work is required to do it).

File: packages/dataviews/src/view-table.js (hunk 1 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: This was me getting ahead of myself, honestly, and it's not necessary; a simple call to `onHide` is more than sufficient at the moment. Will change.

File: packages/dataviews/src/view-table.js (hunk 3 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: Yes, it should probably be something rather than nothing.

File: packages/dataviews/src/view-table.js (hunk 4 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: Yeah, it's less than ideal, and you're right, we shouldn't be relying on incidental behaviour. Will revisit.

File: packages/dataviews/src/view-table.js (hunk 7 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: No reason, wasn't thinking clearly! Will make it clearer.

File: packages/dataviews/src/view-table.js (hunk 2 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: As per @oandregal's suggestion, I've removed these changes in this PR, with a view to normalising them more generally. It doesn't impact focus persistence, so better to free this conversation from that.

File: packages/dataviews/src/view-table.js [hunk 10]
```diff
@@ -366,14 +416,19 @@ function ViewTable( {
...
-			) }
-			{ ! hasData && (
-				<div className="dataviews-no-results">
-					<p>{ __( 'No results' ) }</p>
-				</div>
-			) }
+				</tbody>
+			</table>
+			<div
+				className={ classNames( 'dataviews-table-status', {
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js:422]: I don't see the classes in use for anything. Perhaps we can remove them for now?

File: packages/dataviews/src/view-table.js [hunk 11]
```diff
@@ -283,62 +301,94 @@ function ViewTable( {
 	isLoading = false,
 	deferredRendering,
 } ) {
+	const headerMenuRefs = useRef( new Map() );
+	const [ nextHeaderMenuToFocus, setNextHeaderMenuToFocus ] = useState();
+
+	useEffect( () => {
+		if ( nextHeaderMenuToFocus ) {
+			nextHeaderMenuToFocus.focus();
+			setNextHeaderMenuToFocus();
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js]: Why do we need to update the `nextHeaderMenuToFocus` here? It triggers a second `ViewTable` render, though I don't understand what for.

File: packages/dataviews/src/view-table.js [hunk 12]
```diff
@@ -122,6 +139,7 @@ function HeaderMenu( { field, view, onChangeView } ) {
 					<DropdownMenuItem
 						prefix={ <Icon icon={ unseen } /> }
 						onClick={ () => {
+							onHide?.( field );
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js]: `onHide` is always present, isn't it?

```suggestion
							onHide( field );
```

File: packages/dataviews/src/view-table.js (hunk 11 above)
Comment (oandregal) [reply packages/dataviews/src/view-table.js]:
```suggestion
```

File: packages/dataviews/src/view-table.js [hunk 13]
```diff
@@ -366,14 +416,19 @@ function ViewTable( {
...
-					<p>{ __( 'No results' ) }</p>
-				</div>
-			) }
+				</tbody>
+			</table>
+			<div
+				className={ classNames( 'dataviews-table-status', {
+					'dataviews-loading': isLoading,
+					'dataviews-no-results': ! hasData && ! isLoading,
+				} ) }
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js:437]: I don't see the classes being used for anything. Perhaps we could remove them for now?

```suggestion
```

**Review by oandregal — APPROVED** [review]
This is a great improvement, thanks!

I'm pre-approving, though I left some suggestions. Please, accept or discard as you see fit.

File: packages/dataviews/src/view-table.js (hunk 12 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: Yes. In my head it was optional 🤷

File: packages/dataviews/src/view-table.js (hunk 11 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: My thinking was that this would reduce the chances of an unexpected refocus if something else caused the table to be rerendered. We could short-circuit it a bit – it would still cause an extra rerender, but it would be processed immediately without anything expensive happening...

```js
const headerMenuToFocusRef = useRef( null );
const [ nextHeaderMenuToFocus, setNextHeaderMenuToFocus ] = useState();

useEffect( () => {
	if ( headerMenuToFocusRef.current ) {
		headerMenuToFocusRef.current.focus();
		headerMenuToFocusRef.current = null;
	}
} );

if ( nextHeaderMenuToFocus ) {
	headerMenuToFocusRef.current = nextHeaderMenuToFocus;
	setNextHeaderMenuToFocus();
	return;
}
```

File: packages/dataviews/src/view-table.js (hunk 13 above)
Comment (andrewhayward) [reply packages/dataviews/src/view-table.js:437]: Both `dataviews-loading` and `dataviews-no-results` add a bit of padding...

```css
.dataviews-no-results,
.dataviews-loading {
	padding: 0 $grid-unit-40;
}
```

I'm happy to add `dataviews-table-status` to that list, and keep this static, if that's preferable, but I wanted to keep the classes in there in case anything specific was added to either of them.

Comment (oandregal) [reply packages/dataviews/src/view-table.js:437]: Ah, nice. Searched for them, but didn't see them. I must have done the wrong search.

File: packages/dataviews/src/view-table.js (hunk 11 above)
Comment (oandregal) [reply packages/dataviews/src/view-table.js]: > My thinking was that this would reduce the chances of an unexpected refocus if something else caused the table to be rerendered.

So, this is an attempt to cover against the situation in which the focus is moved elsewhere _after_ it has been moved to the `headerMenuToFocusRef`, correct? I don't see how setting `headerMenuToFocusRef` to `undefined` would help with re-gaining the focus, as `nextHeaderMenuToFocus.focus();` would never be called in that situation. 

Is there a reproducible situation, or did you run into this while testing? I didn't, so my suggestion would be to remove this extra step. But always happy to reconsider.

Comment (andrewhayward) [reply packages/dataviews/src/view-table.js]: If we don't clear `nextHeaderMenuToFocus`, then every time the view is rendered the menu will be focused. For example, if we drop a column, then switch page, focus will be moved to the fallback column header...

https://github.com/WordPress/gutenberg/assets/159848/882aba62-87a1-4be8-af5a-97c666f45b51

Comment (oandregal) [reply packages/dataviews/src/view-table.js]: Ah, so it's the other way around: it's to avoid the header from taking the focus when it has been moved elsewhere. Thanks for the example!

Let's ship as it is and re-consider later.
